org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacRecoveredTypeBinding.java

```diff
@@ -182,4 +230,32 @@ public IVariableBinding[] getDeclaredFields() {
 	public IMethodBinding[] getDeclaredMethods() {
 		return new IMethodBinding[0];
 	}
+	@Override
+	public ITypeBinding getDeclaringClass() {
+		if (this.typeSymbol != null) {
+			return super.getDeclaringClass();
+		}
+		return null;
+	}
+	@Override
+	public String getName(boolean checkParameterized, boolean sourceName) {
+		if (this.typeSymbol != null) {
+			return super.getName(checkParameterized, sourceName);
+		}
+		return this.domNode.toString();
+	}
+	@Override
+	public String getQualifiedName(boolean includeParams) {
+		if (this.typeSymbol != null) {
+			return super.getQualifiedName(includeParams);
+		}
+		return this.domNode.toString();
+	}
+	@Override
+	public String getKey() {
+		if (this.type != null) {
+			return super.getKey();
+		}
+		return "L" + getName() + ";";
```

Comment:
It should probably be `"L" + getQualifiedName() + ";"` The qualified name should include the packageName, which for an unresolved type is expected to be the package of the current unit.